### ISSUE TITLE
At least on windows, configure.py --boostrap fails unless python is in the path

### ISSUE THREAD

**Issue by gmcode** [ninja-build/ninja#1632]
At around line 222 in configure.py, this line exists: 
                  default=os.path.basename(sys.executable))
I'd like to suggest the line be changed to something more robust. This would solve my issue:
                  default=sys.executable)

But I suspect the best change is more involved. Certainly unconditionally just stripping the path away leads to problems, which I'll explain:

On my machine, Python is installed and the .py extension is registered to the path where it is installed to, but python.exe is not on any path - because that's what registration allows for and the python install options offer.

The issue is that in this configuration, typing configure.py --bootstrap at the command line, results in this fail sequence:
1. python starts, because the command line finds python.exe through it's .py extension registration (and not through the path because it's not on the path).
2. python.exe starts the configure.py program as requested and all as expected.
3. configure.py writes a stripped python path consisting of just python.exe and not the full path, when it creates the /DNINJA_PYTHON macro and starts to build ninja.exe.
4. When ninja runs, it tries to run python.exe again, using the stripped path but CreateProcesss in subprocess-win32.cc fails because it python.exe isn't on the path and all it has is a filename to work with because that's all configure.py gave it.

But the filename isn't enough, it needs to have the whole path to run, or just the .py script part without python.exe, so CreateProcess it can use the .py extension to find python. (i.e. in just the same way as the original configure.py --bootstrap command kicked things off.)

The other issue here is that the CreateProcess call in subprocess-win32.cc or any call to it in ninja, doesn't include the filename of what it couldn't run, which made this error tricky to track down in. The command that failed needs to be displayed by configure.py or the filename needs to be in the error.

Simply, not stripping the path solves the  problem. But it's not clear to me that just simply stripping or not stripping the path is the right solution.

As things stand, stripping the path to the filename seems to lead to building a program that's tied to some filename that made sense on someone's machine and it assumes that filename will be the  same and *in the path* on every other machine.

But not stripping and using the full path results in building a ninja tied to a python in a specific location on someone's machine, but at least in the later case it works for the machine that it's built on when python isn't in the path but is installed and registered, whereas the first solution doesn't work and results in the situation I've described in this case.

So, the best solution may need a more involved fix, whatever that is?

A more robust solution, at least for Windows, maybe that in ninja.exe, if the NINJA_PYTHON macro is not defined or is empty or is to a program that is just a filename not a full path and if that filename cannot be found on the PATH, then we interpret this situation to mean the user has been deliberately vague and can't work if tried anyway, so in this case ninja should maybe avoid using the name at all in that case, and just give CreateProcess the raw .py command string without python.exe in it, and leave CreateProcess to do the right thing, meaning it will  go find the .exe associated with .py and go run it.

The above would seem to be more sensible,  more robust, and support path and non path setups of python. We could ignore the python path, if we are given a full python path but it doesn't resolve to an existing program. However in this case, it means we would be deliberately avoiding the users explicit choice (because it doesn't work), to try to find something that does work. That's even more robust, but it could as too generous/try hard.

This logic could go in ninja or configure.py but for robustness it probably needs to be in ninja.

At a minimum, there needs to be an option to run configure.py that says don't strip the path from python so the no path python configuration can be supported. This is an easy change but the user needs to know with clarity when to use this. You likely wouldn't want this to be default because it's use ties the ninja to the path, but the process will blow up with the current situation I am reporting and no error giving the user a clue what to do about it, if it's not the default.

So at a minimum, if you don't change ninja.exe to be more robust, configure.py needs to have an option that says 'hey I see python can't be found on the path so ninja can't be built unless you use the don't strip the path option, but if you do this will tie ninja to this configuration'.

This would support my configuration and tell me when I need it at the cost of tying my ninja.exe to my machine.

If the first option was chosen about making ninja.exe more robust in finding python, then ninja would just automatically support python being in the path or not by trying it if it's there and trying something else if not.

This change may be windows specific. I'm not sure what other operating systems it might be relevant to.

In summary, this change suggested would:
1. enable ninja to support the python configuration option of installing python and registering it's extension but not adding python to any run paths.
2. make ninja more robust and support it's current deployment scenario and another suggested one.
3. improve it's error reporting by including the filename in the error.

Hopefully this gives you enough information to find some solution that enables the scenario I'm interested in fixing, even if it isn't one of the solutions I've suggested.

Thanks

**Comment by jhasse** [comment]
What do you think about using CMake like in #1563?

**Comment by gmcode** [comment]
Using cmake to build ninja instead of python seems much more sensible to me.

**Comment by jhasse** [comment]
I've merged the CMake PR, looking forward to your feedback :)

I hope you're okay with me closing this issue now.
